beacon-chain/mainchain/service.go

```diff
@@ -0,0 +1,100 @@
+package mainchain
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"strings"
+
+	ethereum "github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	gethTypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/rpc"
+	log "github.com/sirupsen/logrus"
+)
+
+// Reader defines a struct that can fetch latest header events from a web3 endpoint.
+type Reader interface {
+	SubscribeNewHead(ctx context.Context, ch chan<- *gethTypes.Header) (ethereum.Subscription, error)
+}
+
+// Web3Service fetches important information about the canonical
+// Ethereum PoW chain via a web3 endpoint using an ethclient. The Random
+// Beacon Chain requires synchronization with the mainchain's current
+// blockhash, block number, and access to logs within the
+// Validator Registration Contract on the mainchain to kick off the beacon
+// chain's validator registration process.
+type Web3Service struct {
+	ctx         context.Context
+	cancel      context.CancelFunc
+	headerChan  chan *gethTypes.Header
+	endpoint    string
+	blockNumber *big.Int    // the latest mainchain blocknumber.
+	blockHash   common.Hash // the latest mainchain blockhash.
+}
+
+// NewWeb3Service sets up a new instance with an ethclient when
+// given a web3 endpoint as a string.
+func NewWeb3Service(endpoint string) (*Web3Service, error) {
+	if !strings.HasPrefix(endpoint, "ws") && !strings.HasPrefix(endpoint, "ipc") {
+		return nil, fmt.Errorf("web3service requires either an IPC or WebSocket endpoint, provided %s", endpoint)
+	}
+	ctx, cancel := context.WithCancel(context.Background())
```

Comment:
This should use a context created by `node` in order to cleanly exit the goroutine in the event the program receives a `SIGTERM`, no?

Reply:
Ah, I see, `node` explicitly calls `Stop` for each service. That should be fine, although deriving the root context for each service (as opposed to creating a new one) might be more efficient once the goroutines get more complicated. Something to keep in mind.